### PR TITLE
Switch between local build and remote build for benchmark tests

### PR DESCRIPTION
**Pull request by gyagp**
Before a public release, we often need to test against local build
instead of npm package, and this change will provide the convenience to
do this.
Usage: index.html?localBuild=<components split by comma>
Example: index.html?localBuild=core,webgl,webgpu

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431)

### CODE REVIEW

**Comment by gyagp** [comment]
@qjia7 @haoyunfeix @xhcao @axinging, PTAL!

File: e2e/benchmarks/local-benchmark/index.html [hunk 1]
```diff
@@ -673,11 +654,68 @@ <h2>TensorFlow.js Model Benchmark</h2>
...
+        }
+        document.body.append(script);
+      })
+    }
+
+    function processUrls(urls, localBuild) {
+      for (let i = 0; i < urls.length; i++) {
+        let name = urls[i].split('/')[0].replace('tfjs-', '').replace('backend-', '');
+        if (localBuild.includes(name)) {
+          urls[i] = '../../../bazel-out/k8-fastbuild/bin/' + urls[i];
```
Comment (gyagp) [inline e2e/benchmarks/local-benchmark/index.html:675]: Currently on Linux, the bazel out dir is called k8-fastbuild, while on Windows, this is called x64_windows-fastbuild. We need to figure out how to change between them. Any suggestion?

Comment (gyagp) [reply e2e/benchmarks/local-benchmark/index.html]: Updated the patch to add another parameter "bazelDir" for this.

**Comment by qjia7** [comment]
@lina128 @pyu10055 @jinjingforever We'd like to hear your opinions about this PR. The main purpose of it is for development to quickly catch any regressions in upstream code instead of waiting for a npm package release. Thanks.

**Review by lina128 — APPROVED** [review]
This is fantastic, thank you for the change! Can you also update the benchmark README too? https://github.com/tensorflow/tfjs/blob/master/e2e/benchmarks/local-benchmark/README.md

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431#-:-MgM9zYa8YSXf8JIdD5j:bebxsdo)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @gyagp, @jinjingforever, and @pyu10055)

___
*[e2e/benchmarks/local-benchmark/index.html, line 675 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5431#-MgF8hO2IYlNhJmSudbm-r1-675:-MgM8Jb52pi-sUN7V0ZR:b2tib92) ([raw file](https://github.com/tensorflow/tfjs/blob/ea4337ccdc1e47b3915e75c06d4524f9e5a83b68/e2e/benchmarks/local-benchmark/index.html#L675)):*
<details><summary><i>Previously, gyagp (Yang Gu) wrote…</i></summary><blockquote>

Updated the patch to add another parameter "bazelDir" for this.
</blockquote></details>

bazelDir is fine, but it does require to know the implementation details. An alternative approach is try ... catch ... Either approach is fine with me.

**Comment by gyagp** [comment]
> This is fantastic, thank you for the change! Can you also update the benchmark README too? https://github.com/tensorflow/tfjs/blob/master/e2e/benchmarks/local-benchmark/README.md

A new section in README is added, and all the known parameters are briefly introduced there. BTW, for consistency, I added numWarmups in the GUI. 

> 
> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431#-:-MgM9zYa8YSXf8JIdD5j:bebxsdo)_ status: :shipit: complete! 1 of 1 approvals obtained (waiting on @gyagp, @jinjingforever, and @pyu10055)
> 
> _[e2e/benchmarks/local-benchmark/index.html, line 675 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5431#-MgF8hO2IYlNhJmSudbm-r1-675:-MgM8Jb52pi-sUN7V0ZR:b2tib92) ([raw file](https://github.com/tensorflow/tfjs/blob/ea4337ccdc1e47b3915e75c06d4524f9e5a83b68/e2e/benchmarks/local-benchmark/index.html#L675)):_
> 
> Previously, gyagp (Yang Gu) wrote…
> bazelDir is fine, but it does require to know the implementation details. An alternative approach is try ... catch ... Either approach is fine with me.

I further checked Bazel source code, and the output dir is in format <cpu>-<compilation_mode>. cpu can be k8, darwin, x64_windows and others, while compilation_mode can be dbg, opt and fastbuild. So the actual name is the combination of all these possibilities, which are too many to enumerate. It seems we have no choice to use bazelDir.

**Review by pyu10055 — CHANGES_REQUESTED** [review]
for bazel directory, I think we make a copy of all the build to tfjs_root/dist/bin
the extra bazel directory might not be needed. @mattsoulanille can you confirm? thanks

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431#-:-Mgd8tnQ4ea2mObpKbZy:b-k2iju6)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @gyagp and @jinjingforever)

___
*[e2e/benchmarks/local-benchmark/index.html, line 696 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5431#-Mgd5zJ2EgwDth8AjfMZ:-Mgd5zJ2EgwDth8AjfM_:b-sqgkta) ([raw file](https://github.com/tensorflow/tfjs/blob/db25ae14a3122b4830a7c7dfd466707fae24e848/e2e/benchmarks/local-benchmark/index.html#L696)):*
> ```HTML
>         'tfjs-backend-cpu/dist/tf-backend-cpu.js',
>         'tfjs-backend-webgl/dist/tf-backend-webgl.js',
>         'tfjs-backend-webgpu/dist/tf-backend-webgpu.min.js',
> ```

should webgpu be consistent with other packages and use non-minimized js file?
___
*[e2e/benchmarks/local-benchmark/README.md, line 68 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5431#-Mgd6WyO2xlh0z1kOYMn:-Mgd6WyO2xlh0z1kOYMo:b-2m15ba) ([raw file](https://github.com/tensorflow/tfjs/blob/db25ae14a3122b4830a7c7dfd466707fae24e848/e2e/benchmarks/local-benchmark/README.md#L68)):*
> ```Markdown
> 
> # Benchmark test
> It's easy to set up a web server to host benchmarks and run against them via e2e/benchmarks/local-benchmark/index.html. Before testing, we need to manually specify all the configurations we'd like to use. For convenience, a bunch of url parameters are supported and listed as below.
> ```

instead of `we need to manually specify all the configurations we'd like to use`
it would be clearer to say `you can manually specify the optional url parameters as needed. Here are the list of supported url parameters:`
___
*[e2e/benchmarks/local-benchmark/README.md, line 76 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5431#-Mgd7EzW4_Fx3nulHw1a:-Mgd7EzW4_Fx3nulHw1b:b-ib58zw) ([raw file](https://github.com/tensorflow/tfjs/blob/db25ae14a3122b4830a7c7dfd466707fae24e848/e2e/benchmarks/local-benchmark/README.md#L76)):*
> ```Markdown
> <b>inputSize</b>: same as inputSizes<br>
> <b>inputType</b>: same as inputTypes<br>
> <b>localBuild</b>: local builds split by comma, which is different from npm packages. Example: 'webgl,core'<br>
> ```

some wording changes:
local build name list, separated by comma. The name is a short form of each npm package (in general the name without the `tfjs-` and `backend-` prefixes, for example `webgl` for `tfjs-backend-webgl`, `core` for `tfjs-core`), Example: 'webgl,core'.

**Comment by gyagp** [comment]
> for bazel directory, I think we make a copy of all the build to tfjs_root/dist/bin
> the extra bazel directory might not be needed. @mattsoulanille can you confirm? thanks
I just confirmed this. Thanks for the info! Done.

> > _[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431#-:-Mgd8tnQ4ea2mObpKbZy:b-k2iju6)_ status: :shipit: complete! 1 of 1 approvals obtained (waiting on @gyagp and @jinjingforever)
> 
> _[e2e/benchmarks/local-benchmark/index.html, line 696 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5431#-Mgd5zJ2EgwDth8AjfMZ:-Mgd5zJ2EgwDth8AjfM_:b-sqgkta) ([raw file](https://github.com/tensorflow/tfjs/blob/db25ae14a3122b4830a7c7dfd466707fae24e848/e2e/benchmarks/local-benchmark/index.html#L696)):_
> 
> > ```
> >         'tfjs-backend-cpu/dist/tf-backend-cpu.js',
> >         'tfjs-backend-webgl/dist/tf-backend-webgl.js',
> >         'tfjs-backend-webgpu/dist/tf-backend-webgpu.min.js',
> > ```
> 
> should webgpu be consistent with other packages and use non-minimized js file?
Due to package size reason, only tf-backend-webgpu.min.js was released. We will switch to the normal one once we totally change the shaders to WGSL.
 
> _[e2e/benchmarks/local-benchmark/README.md, line 68 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5431#-Mgd6WyO2xlh0z1kOYMn:-Mgd6WyO2xlh0z1kOYMo:b-2m15ba) ([raw file](https://github.com/tensorflow/tfjs/blob/db25ae14a3122b4830a7c7dfd466707fae24e848/e2e/benchmarks/local-benchmark/README.md#L68)):_
> 
> > ```gfm
> > # Benchmark test
> > It's easy to set up a web server to host benchmarks and run against them via e2e/benchmarks/local-benchmark/index.html. Before testing, we need to manually specify all the configurations we'd like to use. For convenience, a bunch of url parameters are supported and listed as below.
> > ```
> 
> instead of `we need to manually specify all the configurations we'd like to use`
> it would be clearer to say `you can manually specify the optional url parameters as needed. Here are the list of supported url parameters:`

Done.

> _[e2e/benchmarks/local-benchmark/README.md, line 76 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5431#-Mgd7EzW4_Fx3nulHw1a:-Mgd7EzW4_Fx3nulHw1b:b-ib58zw) ([raw file](https://github.com/tensorflow/tfjs/blob/db25ae14a3122b4830a7c7dfd466707fae24e848/e2e/benchmarks/local-benchmark/README.md#L76)):_
> 
> > ```gfm
> > <b>inputSize</b>: same as inputSizes<br>
> > <b>inputType</b>: same as inputTypes<br>
> > <b>localBuild</b>: local builds split by comma, which is different from npm packages. Example: 'webgl,core'<br>
> > ```
> 
> some wording changes:
> local build name list, separated by comma. The name is a short form of each npm package (in general the name without the `tfjs-` and `backend-` prefixes, for example `webgl` for `tfjs-backend-webgl`, `core` for `tfjs-core`), Example: 'webgl,core'.

Done.

**Review by pyu10055 — APPROVED** [review]
Thank you Yang, this is very useful.

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431#-:-MgfoXk99u1oBs9qZeAb:b-nr84jx)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @gyagp, @jinjingforever, and @pyu10055)

**Review by mattsoulanille — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5431)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @gyagp, @jinjingforever, and @pyu10055)

___
*[e2e/benchmarks/local-benchmark/index.html, line 675 at r1](https://reviewable.io/reviews/tensorflow/tfjs/5431#-MgF8hO2IYlNhJmSudbm-r1-675:-MgkuNtyBMsJBcBeylT3:b5tukti) ([raw file](https://github.com/tensorflow/tfjs/blob/ea4337ccdc1e47b3915e75c06d4524f9e5a83b68/e2e/benchmarks/local-benchmark/index.html#L675)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

bazelDir is fine, but it does require to know the implementation details. An alternative approach is try ... catch ... Either approach is fine with me.
</blockquote></details>

This approach of loading from `dist/bin/` should work on different platforms. If you just need bundles, this works fine, but if you ever need the full npm package, make sure to reference the output of the `pkg_npm` rule. For `tfjs-core`, this would be `dist/bin/tfjs-core/tfjs-core_pkg` (and the bundle would be at `dist/bin/tfjs-core/tfjs-core_pkg/dist/tf-core.js`). The name of the `pkg_npm` directory is defined in each package's BUILD.bazel file.

**Comment by gyagp** [comment]
@mattsoulanille, thanks for the comment! I think we just need bundles now for tests. Please help to merge the PR if you think all the comments are addressed.

**Pull request merged**